src/Database/LSMTree/Internal/MergeSchedule.hs

```diff
@@ -700,24 +713,24 @@ addRunToLevels tr conf@TableConfig{..} resolve hfs hbio root uc r0 reg levels =
             mr' <- newMerge LevelLevelling Merge.LastLevel ln (rs' `V.snoc` r)
             pure $! Level mr' V.empty `V.cons` V.empty
 
+    -- TODO: somehow, this NOINLINE is preventing a NoThunks failure.
+    {-# NOINLINE expectCompletedMerge #-}
     expectCompletedMerge :: LevelNo -> MergingRun m h -> m (Run m h)
     expectCompletedMerge ln (SingleRun r) = do
       traceWith tr $ AtLevel ln $ TraceExpectCompletedMergeSingleRun (runNumber $ Run.runRunFsPaths r)
       pure r
-    expectCompletedMerge ln (MergingRun _ _ var) = do
+    expectCompletedMerge ln (MergingRun _ _ _acVar var) = do
       modifyMVarMasked var $ \case
         x@(CompletedMerge r) -> do
           traceWith tr $ AtLevel ln $ TraceExpectCompletedMerge (runNumber $ Run.runRunFsPaths r)
           pure (x, r)
-        OngoingMerge _rs _ _m -> do
-          -- TODO: should this complete the merge instead, and issue a warning
-          -- (e.g., through a trace message)? It is unclear what is best. On the
-          -- one hand, it is not a case we can not recover from, because we can
-          -- step the merge to completion. On the other hand calling
-          -- expectCompletedMerge on an incomplete merge actually means we have
-          -- a bug in our scheduling, since we take great care to ensure merges
-          -- always finish on time.
-          error "expectCompletedMerge: expected a completed merge, but found an ongoing merge"
+        OngoingMerge rs _ m -> do -- TODO: add warning
+            (_n0, r) <- Merge.stepsToCompletionCounted m 1024
+            RefCount n <- Merge.readRefCount m
+            let !n' = fromIntegralChecked n
+            V.forM_ rs $ \r20 -> Run.removeReferenceN r20 n'
+            Merge.removeReferenceN m n'
+            pure $! (CompletedMerge r, r)
```

Comment:
Note to self

This part should be rewritten a little, so that we can disambiguate between a few cases:
* We're doing some remaining work because we didn't go over the credit threshold
* We're doing some remaining work because we accumulated credits but didn't perform the work
* We're expecting the merge to be finished within this number of credits

Reply:
Done. We do work for the remaining unspent credits. If the merge is not finished then, we error.